### PR TITLE
Prototype of best-chain abstractions

### PR DESCRIPTION
**Pull request by daira**
Definitions for transactions, TXOs, validation contexts, and blocks. Some basic consistency rules are implemented for transactions and blocks, assuming a Bitcoin-like UTXO protocol but with explicit fees. This should be sufficient to require us to take account of UTXO-related issues like missing inputs and double-spends. Both transparent and shielded abstractions are modelled.

Left for future work: it isn't clear whether this modelling of anchors is sufficient, and we might also need to model transaction expiry.

refs #7

### CODE REVIEW

File: simtfl/bc/__init__.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        self.transactions = transactions
+        self.hash = Unique()
+        assert allow_invalid or self.is_valid()
+
+    def is_valid(self):
+        return (
+            len(self.transactions) > 0 and
+            self.transactions[0].is_coinbase() and
+            not any((tx.is_coinbase() for tx in self.transactions[1:])) and
+            sum((tx.fee for tx in self.transactions)) == 0
```
Comment (nathan-at-least) [inline simtfl/bc/__init__.py:91]: Suggestion: Ensure `BCBlock.is_valid` can only be true when every `BCTransaction` is valid (context-freely).

One approach:

1. Define `BCTransaction.is_context_free_valid` to contain the expressions in [the constructor assertions](https://github.com/zcash/simtfl/pull/12/files#diff-6c06bdd8e2d5bf92d00d78832493d2c82444d44a7550b51b56c2d39de66e1c9cR24-R27).
2. Replace those assertion lines with `assert self.is_context_free_valid()`.
3. Add a clause to `BCBlock.is_valid` of: `… and (tx.is_context_free_valid() for tx in self.transactions)`.

File: simtfl/bc/__init__.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        assert ctx.add_if_valid(issuance_tx0)
+        genesis = BCBlock(None, 1, [issuance_tx0])
+        assert genesis.score == 1
+
+        issuance_tx1 = BCTransaction([], [6], -1, issuance=5)
+        spend_tx = BCTransaction([issuance_tx0.output(0)], [9], 1)
+        assert ctx.add_if_valid(issuance_tx1)
+        assert ctx.add_if_valid(spend_tx)
+        block1 = BCBlock(genesis, 1, [issuance_tx1, spend_tx])
+        assert block1.score == 2
```
Comment (nathan-at-least) [inline simtfl/bc/__init__.py:115]: Do we need the simulation to distinguish between validity failure cases? If so, let's add test coverage that ensures these edge cases lead to invalid blocks:

- Blocks:
  - 0 txns
  - non-coinbase initial txn
  - multiple coinbase txns
  - unbalanced fees (note: Doesn't Zcash NU5 mainnet allow unbalanced fees where coinbase claims less than the total allowed? Probably safe to ignore unless there were any possible security impacts for that edge case.)
  - contains "context-free-invalid" txns
  - contains txns with nonexistent (e.g. out-of-`BCContext` inputs)
  - contains txns with multi-spends (e.g. >1 txn with the same input).

File: simtfl/bc/__init__.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        return (
+            len(self.transactions) > 0 and
+            self.transactions[0].is_coinbase() and
+            not any((tx.is_coinbase() for tx in self.transactions[1:])) and
+            sum((tx.fee for tx in self.transactions)) == 0
+        )
+
+
+@dataclass
+class BCProtocol:
```
Comment (nathan-at-least) [inline simtfl/bc/__init__.py:96]: This class doesn't appear to be used. If not, I prefer to omit it and only include used code (even when unused code is anticipating future PRs).

File: simtfl/node.py [hunk 4]
```diff
@@ -56,16 +56,16 @@ def __init__(self, ident, env, network):
         Constructs a SequentialNode with the given simpy Environment and network.
```
Comment (nathan-at-least) [inline simtfl/node.py:56]: I don't understand the motivation for `SequentialNode`, because I assumed `simpy` provides a total ordering of events, so that the order processed by a node will be determinstic.

If my assumption is wrong, then `SequentialNode` makes sense but I would still prefer globally deterministic ordering in order to have globally deterministic simulations, unless we're absolutely certain we need to sacrifice that. (Is performance the only rationale for sacrificing globally deterministic simulation?)

File: simtfl/util.py [hunk 5]
```diff
@@ -8,3 +8,8 @@ def skip():
     """
     # Make this a generator.
     yield from []
+
+
+class Unique:
```
Comment (nathan-at-least) [inline simtfl/util.py:13]: This appears to be unused code. Generally I prefer each PR to contain only used code to keep the overall code as lean as possible. This is because I believe "anticipatory code" often needs to be modified once used, and worse it can sometimes accrue while remaining unused unintentionally, leading to unnecessary complexity.

**Review by nathan-at-least — CHANGES_REQUESTED** [review]
There's at least one comment that is either a bug or naming/design-confusion. Most other comments are suggestions but not requirements for merge.

File: .flake8 [hunk 6]
```diff
@@ -1,4 +1,4 @@
 [flake8]
 exclude = .git, __pycache__
-ignore = E302
+ignore = E302, W503, W504
```
[the remark this answers is not in the document]
Comment (daira) [reply .flake8:3]: A similar change is in #18.

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+on: pull_request
+
+jobs:
+  verify:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Install gnome-keyring
```
[the remark this answers is not in the document]
Comment (daira) [reply .github/workflows/tests.yml:12]: It's copied from zcash-test-vectors which also uses poetry. @nuttycom added it here: https://github.com/zcash-hackworks/zcash-test-vectors/commit/73ced0f5c986d7dceeb231ce5bf1822bb6dc2673

File: simtfl/bc/__init__.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        """
+        Constructs a `BCTransaction` with the given inputs, output values, fee,
+        and (if it is a coinbase transaction) issuance.
+        For a coinbase transaction, pass `inputs=[]`, and `fee` as a negative value
+        of magnitude equal to the total amount of fees paid by other transactions
+        in the block.
+        """
+        assert issuance >= 0
+        assert fee >= 0 or len(inputs) == 0
+        assert issuance == 0 or len(inputs) == 0
```
Comment (daira) [inline simtfl/bc/__init__.py:26]:
```suggestion
        assert issuance == 0 or len(inputs) == 0
        assert all((isinstance(txin, self._TXO) for txin in inputs))
```

File: simtfl/bc/__init__.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+class BCContext:
+    def __init__(self):
+        """Constructs an empty `BCContext`."""
+        self.transactions = deque()
+        self.utxo_set = set()
+        self.total_issuance = 0
+
+    def is_valid(self, tx):
+        """Is `tx` valid in this context?"""
+        return set(tx.inputs).issubset(self.utxo_set)
```
[the remark this answers is not in the document]
Comment (daira) [reply simtfl/bc/__init__.py:55]: `BCTransaction._TXO`, yes. We don't know a priori whether they're unspent; that's what we're checking here (for this context).

File: simtfl/bc/__init__.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+    def add_if_valid(self, tx):
+        """
+        If `tx` is valid in this context, add it to the context and return `True`.
+        Otherwise leave the context unchanged and return `False`.
+        """
+        txins = set(tx.inputs)
+        valid = txins.issubset(self.utxo_set)
+        if valid:
+            self.utxo_set -= txins
```
[the remark this answers is not in the document]
Comment (daira) [reply simtfl/bc/__init__.py:65]: This converts the inputs to a set twice, which is an O(n log n) operation. I know we're not too worried about performance, but this seemed as though it could potentially be measurable. The code that is being duplicated is a single line a few lines apart, so it seems fine.

File: simtfl/bc/__init__.py (hunk 1 above)
Comment (daira) [reply simtfl/bc/__init__.py:91]: I intentionally want to be able to construct invalid blocks for attacks. Also, a given block's transactions can be (and are, in Snap-and-Chat or Crosslink) contextually validated in multiple contexts. So it's necessary to separate non-contextual from contextual validity (as `zcashd` and `zebra` do in practice), and cache whether non-contextual validity has been checked. In order to do that, `is_valid` *only* tests non-contextual validity (but this should be documented).

File: simtfl/bc/__init__.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        self.hash = Unique()
+        assert allow_invalid or self.is_valid()
+
+    def is_valid(self):
+        return (
+            len(self.transactions) > 0 and
+            self.transactions[0].is_coinbase() and
+            not any((tx.is_coinbase() for tx in self.transactions[1:])) and
+            sum((tx.fee for tx in self.transactions)) == 0
+        )
```
[the remark this answers is not in the document]
Comment (daira) [reply simtfl/bc/__init__.py:92]: It's not a bug. I'll rename the method.

File: simtfl/bc/__init__.py (hunk 3 above)
Comment (daira) [reply simtfl/bc/__init__.py:96]: I know it will be used soon, so it's just make-work to change this.

File: simtfl/bc/__init__.py (hunk 2 above)
Comment (daira) [reply simtfl/bc/__init__.py:115]: Yes, I intend to improve the test coverage later.

File: simtfl/util.py (hunk 5 above)
Comment (daira) [reply simtfl/util.py:14]: It's not unused; the `hash` value of a `BCBlock` is set to a `Unique` value. Again, I know this will be used, and being careful to make sure that the boundaries of PRs are such that everything is used in each PR is IMHO just make-work.

(If we were programming a production system then I'd agree with you; for prototyping, I don't. We're already doing this *way* more formally than is typical for prototyping.)

File: simtfl/bc/__init__.py (hunk 1 above)
Comment (daira) [reply simtfl/bc/__init__.py:93]: On reflection I don't think this provides any benefit. Since you can't construct a `BCTransaction` other than via the constructor, it is already the case that you can't construct a `BCBlock` with non-contextually invalid transactions, except by committing a type error or passing a subclass that violates the intended invariants.

File: simtfl/bc/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (daira) [reply simtfl/bc/__init__.py:111]: Renamed to `is_noncontextually_valid`.

File: .flake8 (hunk 6 above)
[the remark this answers is not in the document]
Comment (daira) [reply .flake8]: The reason why we suppress both W503 and W504, is that I think it's an outright bug in `flake8` that either of these warnings trigger on cases where you've intentionally spread code across lines by leaving a parenthesis open. The whole point of doing that is *so that* you can put operators at the beginning or end of a line. Whether the code reads more clearly with the operator at the beginning or end can depend on which operator it is.

File: simtfl/node.py (hunk 4 above)
Comment (daira) [reply simtfl/node.py]: The ordering is always deterministic (if you don't write explicitly nondeterministic code). The difference between `PassiveNode` and `SequentialNode` is that `PassiveNode` doesn't block handling the next message on finishing handling this one, and `SequentialNode` does.

It's likely to be easier to write nodes with sequential message handling. However, `PassiveNode` has a simpler implementation, and concurrent message handling is sometimes necessary.

Comment (nathan-at-least) [reply simtfl/node.py]: Ah, I understand now. I missed that difference but re-reviewing `SequentialNode.run` I now see how it does not process a pending message from `self._mailbox` until all of `yield from self.handle(sender, message)` completes.

File: .github/workflows/tests.yml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+        run: sudo apt-get install gnome-keyring
```
Comment (nathan-at-least) [inline .github/workflows/tests.yml:13]: I request we remove this step if its not needed, otherwise add a comment about why its needed.

I'm allergic to unnecessary changes getting in, then persisting through legacy, because as those build up it makes it more difficult for newcomers to filter through that kind of distraction.

**Review by nathan-at-least — DISMISSED** [review]
One request to delete a CI step based on the guess that it's unnecessary (because I'm allergic to unnecessary code). If it is necessary, I request we add a comment as to why (for collaborators like me where it's not obvious; it may be obvious for anyone with more github CI experience).

File: .github/workflows/tests.yml (hunk 12 above)
Comment (daira) [reply .github/workflows/tests.yml:14]: Comment added.

File: simtfl/util.py (hunk 5 above)
Comment (daira) [reply simtfl/util.py:14]: It is nontrivially used now.

File: .github/workflows/tests.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (daira) [reply .github/workflows/tests.yml]: A comment was added for this in #18.

File: simtfl/bc/__init__.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (daira) [reply simtfl/bc/__init__.py]: This has been improved to both not duplicate code, and not convert the inputs to a set twice.
